aws/logs_monitoring/parsing.py

```diff
@@ -528,6 +528,119 @@ def cwevent_handler(event, metadata):
     yield data
 
 
+def parse_aws_waf_logs(event):
+    """Parse out complex arrays of objects in AWS WAF logs
+
+    Attributes to convert:
+        httpRequest.headers
+        nonTerminatingMatchingRules
+        rateBasedRuleList
+        ruleGroupList
+
+    This prevents having an unparsable array of objects in the final log.
+    """
+    if event.get(DD_SOURCE) != "waf":
+        return event
+
+    event_copy = copy.deepcopy(event)
+
+    message = event_copy.get("message", {})
+    if isinstance(message, str):
+        try:
+            message = json.loads(message)
+        except json.JSONDecodeError:
+            logger.debug("Failed to decode waf message")
+            return
+
+    headers = message.get("httpRequest", {}).get("headers")
+    if headers and isinstance(headers, list):
+        header_obj = {}
+        for header in headers:
+            if "name" in header and "value" in header:
+                header_obj.update({header["name"]: header["value"]})
+        message["httpRequest"]["headers"] = header_obj
+
+    # Iterate through rules in ruleGroupList and nest them under the group id
+    # ruleGroupList has three attributes that need to be handled separately
+    rule_groups = message.get("ruleGroupList", {})
+    if rule_groups and isinstance(rule_groups, list):
+        message["ruleGroupList"] = {}
+        for rule_group in rule_groups:
+            group_id = None
+            if "ruleGroupId" in rule_group and rule_group["ruleGroupId"]:
+                group_id = rule_group.pop("ruleGroupId", None)
+            if not group_id in message["ruleGroupList"]:
+                message["ruleGroupList"][group_id] = {}
+
+            # Extract the terminating rule and nest it under its own id
```

Comment:
Let's refactor these blocks into separate functions to reduce the length & complexity of the `parse_aws_waf_logs` function.